arrow-libs/core/arrow-core/src/commonMain/kotlin/arrow/core/continuations/EagerEffectScope.kt

```diff
@@ -186,6 +188,56 @@ public interface EagerEffectScope<in R> {
    */
   public suspend fun ensure(condition: Boolean, shift: () -> R): Unit =
     if (condition) Unit else shift(shift())
+
+  /**
+   * Encloses an action for which you want to catch any `shift`.
+   * [attempt] is used in combination with [catch].
+   *
+   * ```
+   * attempt { ... } catch { ... }
+   * ```
+   *
+   * The [f] may `shift` into a different `EagerEffectScope`, giving
+   * the chance for a later [catch] to change the shifted value.
+   * This is useful to simulate re-throwing of exceptions.
+   */
+  @OptIn(ExperimentalTypeInference::class)
+  public suspend fun <E, A> attempt(
+    @BuilderInference
+    f: suspend EagerEffectScope<E>.() -> A,
+  ): suspend EagerEffectScope<E>.() -> A = f
+
+
+  /**
+   * When the [EagerEffect] has shifted with [R] it will [recover]
+   * the shifted value to [A], and when it ran the computation to
+   * completion it will return the value [A].
+   * [catch] is used in combination with [attempt].
+   *
+   * ```kotlin
+   * import arrow.core.Either
+   * import arrow.core.None
+   * import arrow.core.Option
+   * import arrow.core.Validated
+   * import arrow.core.continuations.eagerEffect
+   * import io.kotest.assertions.fail
+   * import io.kotest.matchers.shouldBe
+   *
+   * fun main() {
+   *   eagerEffect<String, Int> {
+   *     val x = Either.Right(1).bind()
+   *     val y = Validated.Valid(2).bind()
+   *     val z =
+   *      attempt { None.bind { "Option was empty" } } catch { 0 }
+   *     x + y + z
+   *   }.fold({ fail("Shift can never be the result") }, { it shouldBe 3 })
+   * }
+   * ```
+   * <!--- KNIT example-eager-effect-scope-08.kt -->
+   */
+  public infix fun <E, A> (suspend EagerEffectScope<E>.() -> A).catch(
+    recover: EagerEffectScope<R>.(E) -> A,
```

Comment:
```suggestion
    recover: suspend EagerEffectScope<R>.(E) -> A,
```

shift in EagerEffect here is also suspend

Reply:
Thanks for catching this!

Reply:
It seems that it doesn't work with `suspend` here. I've tried adding `suspend` to the top-level too, but I cannot make it work, so I've reverted it. I think this is fine, even though we might find a more general `suspend`-based version in the future.